BootloaderCommonPkg/Library/ShellLib/CmdUsbDev.c

```diff
@@ -66,55 +66,66 @@ ShellCommandUsbDevFunc (
   UINTN              StringSize;
   UINTN              StringIndex;
   UINTN              CharFound;
+  UINT16             DeviceCounter;
+  PLT_DEVICE_TABLE   *DeviceTable;
+  PLT_DEVICE         *Device;
+  UINTN              DeviceIndexNumber;
 
-  BaseAddress = GetDeviceAddr (OsBootDeviceUsb, 0);
-  if (BaseAddress == 0) {
-    ShellPrint (L"Base Address is 0");
-    return EFI_ABORTED;
-  } else if (!(BaseAddress & 0xFF000000)) {
-    BaseAddress = TO_MM_PCI_ADDRESS (BaseAddress);
-  }
+  DeviceIndexNumber = 1;
+  DeviceTable = (PLT_DEVICE_TABLE *)GetDeviceTable();
 
-  Status = InitUsbDevices (BaseAddress);
-  if (!EFI_ERROR(Status)) {
-    Status = GetUsbDevices ((PEI_USB_IO_PPI **)&UsbIoArray, &UsbIoCount);
-  }
-  if (EFI_ERROR(Status)) {
-    ShellPrint (L"Failed to initialize USB bus !\n");
-    return Status;
-  }
+  for(DeviceCounter = 0; DeviceCounter < DeviceTable->DeviceNumber; DeviceCounter++) {
+    Device = &DeviceTable->Device[DeviceCounter];
 
-  ShellPrint (L"\n");
-  for (Index = 0; Index < UsbIoCount; Index++) {
-    NameStr = GetUsbDeviceNameString (UsbIoArray[Index]);
-    if (NameStr == NULL) {
-      NameStr = L"N/A";
+    if (Device->Type != OsBootDeviceUsb) {
+      continue;
     }
-    StringSize = StrSize (NameStr);
 
-    CharFound = 0;
-    for (StringIndex = 0; StringIndex < StringSize; StringIndex++) {
-      if (NameStr[StringIndex] == 0) {
-        break;
+    BaseAddress = Device->Dev.DevAddr;
+    if (!(BaseAddress & 0xFF000000)) {
+      BaseAddress = TO_MM_PCI_ADDRESS (BaseAddress);
+    }
+
+    Status = InitUsbDevices (BaseAddress);
+    if (!EFI_ERROR(Status)) {
+      Status = GetUsbDevices ((PEI_USB_IO_PPI **)&UsbIoArray, &UsbIoCount);
+    }
+    if (EFI_ERROR(Status)) {
+      ShellPrint (L"Failed to initialize USB bus !\n");
+      return Status;
+    }
+
+
+    for (Index = 0; Index < UsbIoCount; Index++) {
+      NameStr = GetUsbDeviceNameString (UsbIoArray[Index]);
+      if (NameStr == NULL) {
+        NameStr = L"N/A";
       }
-      //
-      // Checking the NameStr is not combination of spaces and tabs
-      //
-      if (((char)NameStr[StringIndex]) != 32 && ((char)NameStr[StringIndex]) != 9) {
-        CharFound++;
+      StringSize = StrSize (NameStr);
+
+      CharFound = 0;
+      for (StringIndex = 0; StringIndex < StringSize; StringIndex++) {
+        if (NameStr[StringIndex] == 0) {
```

Comment:
It looks this check is not necessary since it could not be 0 before String Size.

Reply:
This check is not to validate the length of the string, it is there to validate those devices' names returned as null strings. If the string is null then it will be marked as an 'Unnamed Device'.

Reply:
If the name string is null, it will be set to "N/A" in line 102 already. so it is not possible to have NULL string here.